clients/src/main/java/org/apache/kafka/common/errors/MemberIdMismatchException.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+public class MemberIdMismatchException extends ApiException {
```

Comment:
Hey Stanis, once I started the implementation, I realized that it's more clear to use the current member.id for static members instead of generating a new one, since we need member.id to track heartbeat & stuffs. I will update the KIP to reflect this change.

As for the `MemberIdMismatchException`, I put the explanation within Errors.java as error message to feedback end user.

Reply:
@abbccdda - judging by `GroupMetadata#replace()`, the current behavior is to generate a new member ID and return it, right?
What happens if a misconfigured consumer joins with an existing, duplicate `consumer.instance.id`? It essentially kicks out the old consumer using it (by invalidating its member.id)? Does the old consumer try to rejoin with its group.instance.id afterwards? We could get into a bad loop if that is the case

Reply:
That’s a very good question. Previously my thought was to use conflict member.id to shut down duplicate consumer instances. However, this probably won’t work because upon receiving UNKNOWN_MEMBER_ID exception in either `SyncGroup, Heartbeat, OffsetCommit` requests will immediately reset the generation info which includes the member.id. One approach I could think of is to restrict the caller of `resetGeneration` on client side to only JoinGroup logic, which means for any other types of requests after receiving UNKNOWN_MEMBER_ID will be rejoining the group with their current generation info (the conflicting member.id). This should be able to help us detect the id collision and shut down duplicate member with MEMBER_ID_MISMATCH exception.

@stanislavkozlovski @hachikuji @guozhangwang Thoughts？

Reply:
I don't understand, who would receive the UNKNOWN_MEMBER_ID?

If consumer A has `member.id=1, instance.id=one` and consumer B joins with `member.id=2, instance.id=one`, wouldn't A receive MEMBER_ID_MISMATCH and shut down?

Reply:
@stanislavkozlovski this won't happen automatically. The flow is like: 
1. consumer A with `member.id=1, instance.id=one` is working under stable group. The static member metadata map contains kv entry `one=1`.
2. consumer B starts up, joining with same `instance.id=one` and `member.id=unknown`
3. consumer B enters `doUnknownJoinGroup` block and successfully gets identity `member.id=2`. The static member metadata map now updates to `one=2`
4. consumer A gets fenced by either `SyncGroup, Heartbeat, OffsetCommit` which informs A with `UNKNOWN_MEMBER_ID` error, which will trigger `resetGeneration()` on client side AbstractCoordinator.
5. Now consumer A rejoins with `instance.id=one` and `member.id=unknown`, repeating step 2 like B.

So eventually A, B will bounce forever within the loop 2~5 unless one of them refuses to reset their assigned member.id. Otherwise `MEMBER_ID_MISMATCH` shall never trigger.

Reply:
Aha, yeah. We can only raise `MEMBER_ID_MISMATCH ` in the `JoinGroup` request because that's the only request that has the group instance id field, right?

As you proposed, I think making the consumer issue a new JoinGroup with the same member.id would be the better approach.  Otherwise, we'd probably need to add the new field to all the requests. The old functionality of resetting the generation should continue to work just fine, we'd just be adding an extra hop.

Reply:
I think restricting `resetGeneration` to only JoinGroup request is not the best approach since we do rely on, e.g. heartbeat response to notify consumers as early as possible. On the other hand, this issue would only raise if users mis configure their `instance.id` to have two running instances to have the same id, such issue is similar to producer client that two instances mistakenly configured with the same `transactional.id` and today it is handled by letting one of them to receive a fatal error (`fenced`) and either handle it themselves or die hard -- the bottom line is, brokers would not need to be responsible for abstracting such human errors from clients.

So I'd like to present an alternative proposal: 

1) when receiving a join group of null member.id, but existing instance.id, create a new member.id just instead of returning the associated member.id to the client (your PR already did this anyways)

2) when receiving a join group of non-empty member.id, and existing instance.id, BUT is inconsistent with the static members map, return error `MEMBER_ID_REQUIRED`.

Now the only issue is what if two instances come with the same instance.id and the same member.id. I think it would not be possible for new members due to 1) since we always generate a new member.id.

Does that make sense to you? cc @stanislavkozlovski @hachikuji